### DISCUSSION TITLE
Cake.Recipe prerelease deployment fails on rebased branch

### DISCUSSION THREAD

**Discussion by eoehen** [cake-build/cake#3200] [Extension Q&A] [answered]
**ENVIRONMENT**
Repository: https://github.com/eoehen/arguard
Build: https://github.com/eoehen/arguard/runs/1955874389
Plattform: Ubuntu
Cake.Recipe Version: 2.2.0

**PROBLEM**
I have one Github Actions CI build workflow for all pushes.
When I rebase a branch the build fails on prerelease deployment, because the generated gitversion prerelease version number is not unique and already exists on target system.

This is a rebased `dependabot` pullrequest:

![image](https://user-images.githubusercontent.com/49316770/109119448-feb9c100-7744-11eb-91e0-8f216d346d4f.png)

**WORKAROUND**
Possible workaround is to create a own GitHub Actions build only for pullrequest builds without prerelease deployment.

**QUESTION**
Can I improve the `GitVersion` configuration, that the generated version number is also unique when I rebase a branch?

**POSSIBLE CHANGE REQUEST IDEAS**
A) A flag parameter to prevent prerelease deployments for pullrequest builds.
B) We never deploy prereleases on pullrequest builds.
C) We add a list of branch name parameter (master, develop, ...) where prereleases (or general all deployment tasks) are executed.

If one of my requests makes sense, I would be interested in implementing it.
Many thanks for feedback.

**Comment by AdmiringWorm** [comment, marked as answer]
This is unfortunately a known problem, and there is no way to configure gitversion in cake.recipe (other than the yaml configuration) without overriding the whole setup (see the documentation here: https://cake-contrib.github.io/Cake.Recipe/docs/extending/override-generating-version).

> A) A flag parameter to prevent prerelease deployments for pullrequest builds.

Not exactly the same, but in an upcoming version of Cake.Recipe this is semi-added.
A recent pull request added the ability to set a new variable called `SkipDuplicatePackages` to the ToolSettings class that allows builds to continue even if a push of NuGet packages fail due to them already existing upstream.

> B) We never deploy prereleases on pullrequest builds.

Cake.Recipe only pushes packages when the necessary environment variables are available, unfortunately due to the pull request coming from a local branch they also are.
Also want to mention that the failure you are seeing is on the branch itself, not on the pull request, and this is due to the way that GitHub Actions is set up, both a build is started on the pull request and on the originating branch when it is in the same repository.
It is this latter build that fails in your case, and there is unfortunately no way for Cake.Recipe to know if that branch is part of a pull request (due to not enough information in GitHub Actions).

> C) We add a list of branch name parameter (master, develop, ...) where prereleases (or general all deployment tasks) are executed.

TBH, I am not sure if this would make sense. Only in some cases.

**Reply by eoehen** [reply]
Thank you very much for your answer.
